diff --git a/CMake/HPHPSetup.cmake b/CMake/HPHPSetup.cmake
index 351db73..d73bf11 100644
--- a/CMake/HPHPSetup.cmake
+++ b/CMake/HPHPSetup.cmake
@@ -115,7 +115,7 @@ IF($ENV{CXX} MATCHES "icpc")
 	set(CMAKE_CXX_FLAGS "-no-ipo -fp-model precise -wd584 -wd1418 -wd1918 -wd383 -wd869 -wd981 -wd424 -wd1419 -wd444 -wd271 -wd2259 -wd1572 -wd1599 -wd82 -wd177 -wd593 -fno-omit-frame-pointer -ftemplate-depth-60 -Wall -Woverloaded-virtual -Wno-deprecated -w1 -Wno-strict-aliasing -Wno-write-strings -Wno-invalid-offsetof -fno-operator-names")
 else()
 	set(CMAKE_C_FLAGS "-w")
-	set(CMAKE_CXX_FLAGS "-fno-gcse -fno-omit-frame-pointer -ftemplate-depth-60 -Wall -Woverloaded-virtual -Wno-deprecated -Wno-strict-aliasing -Wno-write-strings -Wno-invalid-offsetof -fno-operator-names -Wno-error=array-bounds -Wno-error=switch -std=gnu++0x -Werror=format-security -Wno-unused-result -Wno-sign-compare")
+	set(CMAKE_CXX_FLAGS "-fno-gcse -fno-omit-frame-pointer -ftemplate-depth-128 -Wall -Woverloaded-virtual -Wno-deprecated -Wno-strict-aliasing -Wno-write-strings -Wno-invalid-offsetof -fno-operator-names -Wno-error=array-bounds -Wno-error=switch -std=gnu++0x -Werror=format-security -Wno-unused-result -Wno-sign-compare")
 endif()
 
 IF(CMAKE_COMPILER_IS_GNUCC)
diff --git a/hphp/legacy/run.sh b/hphp/legacy/run.sh
index bede609..cc75d66 100755
--- a/hphp/legacy/run.sh
+++ b/hphp/legacy/run.sh
@@ -5,6 +5,15 @@
    #echo make -j $3 PROJECT_NAME=$2 TIME_LINK=1 -C $1
 cp $HPHP_HOME/hphp/legacy/CMakeLists.base.txt $1/CMakeLists.txt
 cd $1
+
+# If hphc was called from a makefile, don't propogate
+# environment nonsense to the child, because this breaks tests
+# I wish there was an exhaustive list of these somewhere
+unset MAKEFLAGS
+unset MAKEOVERRIDES
+unset MFLAGS
+unset MAKELEVEL
+
 cmake -D PROGRAM_NAME:string=$2 . || exit $?
 
 if [ -n "$HPHP_VERBOSE" ]; then
diff --git a/hphp/rules.mk b/hphp/rules.mk
index e6f53e1..6375aa0 100644
--- a/hphp/rules.mk
+++ b/hphp/rules.mk
@@ -254,7 +254,7 @@ GOOGLE_TOOLS = 1
 endif
 
 CPPFLAGS += -D_GNU_SOURCE -D_REENTRANT=1 -D_PTHREADS=1 -pthread
-CXXFLAGS += -ftemplate-depth-60
+CXXFLAGS += -ftemplate-depth-128
 
 endif
 
diff --git a/hphp/runtime/base/type_string.cpp b/hphp/runtime/base/type_string.cpp
index 0290872..fc25f4f 100644
--- a/hphp/runtime/base/type_string.cpp
+++ b/hphp/runtime/base/type_string.cpp
@@ -41,8 +41,6 @@ const StaticString empty_string("");
 StringData const **String::converted_integers_raw;
 StringData const **String::converted_integers;
 
-String::IntegerStringDataMap String::integer_string_data_map;
-
 static const StringData *convert_integer_helper(int64 n) {
   char tmpbuf[21];
   char *p;
@@ -55,6 +53,7 @@ static const StringData *convert_integer_helper(int64 n) {
 }
 
 void String::PreConvertInteger(int64 n) {
+  IntegerStringDataMap& integer_string_data_map = GetIntegerStringDataMap();
   IntegerStringDataMap::const_iterator it =
     integer_string_data_map.find(n);
   if (it != integer_string_data_map.end()) return;
diff --git a/hphp/runtime/base/type_string.h b/hphp/runtime/base/type_string.h
index a69d0b8..e5dd7ac 100644
--- a/hphp/runtime/base/type_string.h
+++ b/hphp/runtime/base/type_string.h
@@ -60,7 +60,13 @@ public:
   static const int MaxPrecomputedInteger = 4095 + SCHAR_MIN;
   static StringData const **converted_integers_raw;
   static StringData const **converted_integers;
-  static IntegerStringDataMap integer_string_data_map;
+
+  // see for details:
+  // https://github.com/facebook/hhvm/issues/570#issuecomment-8831861
+  static IntegerStringDataMap& GetIntegerStringDataMap() {
+    static IntegerStringDataMap integer_string_data_map;
+    return integer_string_data_map;
+  }
 
   static bool HasConverted(int64 n) {
     return MinPrecomputedInteger <= n && n <= MaxPrecomputedInteger;
@@ -84,6 +90,7 @@ public:
       }
       return sd;
     }
+    IntegerStringDataMap& integer_string_data_map = GetIntegerStringDataMap();
     IntegerStringDataMap::const_iterator it =
       integer_string_data_map.find(n);
     if (it != integer_string_data_map.end()) return it->second;
diff --git a/hphp/runtime/tmp/test.mk b/hphp/runtime/tmp/test.mk
index 1ad42c8..3e903e6 100644
--- a/hphp/runtime/tmp/test.mk
+++ b/hphp/runtime/tmp/test.mk
@@ -19,7 +19,7 @@ TARGET = hphp
 EXTRA_RUNTIME_OPTIONS:=-vEval.Jit=$(if $(filter TestCodeRunRepoJit,$(SUITE)),1,0)
 endif
 
-ZEND = /home/engshare/externals/cpp/hphp/centos-dev/php/bin/php
+ZEND = php
 ZEND_WARN = -ddisplay_errors=stderr -dapc.enable_cli=1
 ZEND_NOWARN = -ddisplay_errors=off -dapc.enable_cli=1
 
diff --git a/hphp/test/test_memcached_info.inc b/hphp/test/test_memcached_info.inc
index c36489a..603529e 100644
--- a/hphp/test/test_memcached_info.inc
+++ b/hphp/test/test_memcached_info.inc
@@ -1,3 +1,3 @@
 // Fill the following macros with hostname and port of memcached
-#define TEST_MEMCACHED_HOSTNAME ""
+#define TEST_MEMCACHED_HOSTNAME "localhost"
 #define TEST_MEMCACHED_PORT     11211
